sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamFilterRel.java

```diff
@@ -45,6 +48,14 @@ public Filter copy(RelTraitSet traitSet, RelNode input, RexNode condition) {
     return new BeamFilterRel(getCluster(), traitSet, input, condition);
   }
 
+  @Override
+  public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    double rowCnt = mq.getRowCount(this);
+    return planner
+        .getCostFactory()
+        .makeCost(rowCnt, rowCnt, rowCnt * estimateRowSize(getRowType()));
```

Comment:
So just checking - this overrides the default of

 - rowcount
 - cpu proportional to rowcount
 - IO zero

and makes it

 - rowcount
 - cpu proportional to rowcount
 - IO proportional to rowcount * schema size

That seems like a good change. Since it is just a vague generalization, though, I think that you might just put IO cost on GBK (since it usually is a shuffle) and sinks, but leave it zero for filters and projects, etc.

Later we'll need to be able to plug in cost models.

Reply:
yes, we can support CBO later. It seems not quite relevant here. I remove it from this patch.